### PR TITLE
feat: add put bso and post collection validator/extractors

### LINKED ISSUE #50: Port put/post validators
Port the put validators:
```
PUT_VALIDATORS = DEFAULT_VALIDATORS + (
    parse_single_bso,
    check_for_known_bad_payloads
)
```

### PR DESCRIPTION
**Pull request by bbangert**
Closes #50

### CODE REVIEW

File: src/web/extractors.rs [hunk 1]
```diff
@@ -33,26 +47,214 @@ pub struct UidParam {
...
+                        let err: ApiError = ValidationErrorKind::FromValidationErrors(
+                            e,
+                            RequestErrorLocation::Body,
+                        ).into();
+                        return future::err(err.into());
+                    }
+                }
+
+                // Validate the TTL, comment copied from Python, bug seems resolved
+                // so this workaround should no longer be necessary, verify during PR review.
```
Comment (philbooth) [inline src/web/extractors.rs]: Question for @rfk...?

Comment (bbangert) [reply src/web/extractors.rs]: I talked with @rfk, he thinks it should be fine to validate it properly now, so I'll update and remove this exception.

File: src/web/extractors.rs [hunk 2]
```diff
@@ -33,26 +47,214 @@ pub struct UidParam {
...
+        let fut = <Json<BsoBody>>::from_request(req, &config)
+            .map_err(|e| {
+                let err: ApiError = ValidationErrorKind::FromDetails(
+                    e.to_string(),
+                    RequestErrorLocation::Body,
+                    Some("bso".to_owned()),
+                ).into();
+                err.into()
+            }).and_then(|mut bso: Json<BsoBody>| {
+                match bso.validate() {
```
Comment (pjenvey) [inline src/web/extractors.rs]: could save a couple lines here w/ if let Err(e) = bso.validate() {

File: src/web/extractors.rs [hunk 3]
```diff
@@ -168,6 +368,81 @@ impl FromRequest<ServerState> for CollectionRequest {
...
+            HawkIdentifier,
+            State<ServerState>,
+            CollectionParam,
+            BsoQueryParams,
+            BsoBodies,
+        )>::extract(req).and_then(|(user_id, state, collection, query, mut bsos)| {
+            let collection = collection.collection.clone();
+            if collection == "crypto" {
+                // Verify the client didn't mess up the crypto if we have a payload
+                for bso in bsos.valid.iter() {
```
Comment (pjenvey) [inline src/web/extractors.rs]: similar here (could be for `bso in &bsos.valid`)

File: src/web/extractors.rs [hunk 4]
```diff
@@ -168,6 +368,81 @@ impl FromRequest<ServerState> for CollectionRequest {
...
+                            );
+                        }
+                    }
+                }
+            }
+
+            // Trim the excess BSO's to be under the batch size
+            let overage: i64 = (bsos.valid.len() as i64) - (BATCH_MAX_RECORDS as i64);
+            if overage > 0 {
+                for _ in 1..=overage {
```
Comment (pjenvey) [inline src/web/extractors.rs]: it probably doesn't matter much but you could use Vec::drain here instead while maintaining the original order of the invalids

```rust
                bsos.invalid
                    .extend(bsos.valid.drain(overage as usize..).map(|last| BsoBody {
                        id: Some(last.id),
                        sortindex: last.sortindex,
                        payload: last.payload,
                        ttl: last.ttl,
                    }));
```

File: src/web/extractors.rs [hunk 5]
```diff
@@ -465,6 +795,31 @@ fn validate_qs_ids(ids: &Vec<String>) -> Result<(), ValidationError> {
     Ok(())
 }
 
+/// Verifies the BSO sortindex is in the valid range
+fn validate_body_bso_sortindex(sort: i32) -> Result<(), ValidationError> {
+    if BSO_MIN_SORTINDEX_VALUE < sort && sort < BSO_MAX_SORTINDEX_VALUE {
```
Comment (pjenvey) [inline src/web/extractors.rs]: you want min <= sort && sort <= max

File: src/web/handlers.rs [hunk 6]
```diff
@@ -142,22 +134,15 @@ pub fn get_bso(bso_req: BsoRequest) -> FutureResponse<HttpResponse> {
...
-            collection: "tabs".to_owned(),
-            id: params.bso.clone(),
-            sortindex: body.sortindex,
-            payload: body.payload.clone(), // XXX: need clone here from body, sigh
-            ttl: body.ttl,
+            user_id: bso_req.user_id,
+            collection: bso_req.collection,
+            id: bso_req.bso.clone(),
+            sortindex: bso_req.body.sortindex,
+            payload: bso_req.body.payload.clone(), // XXX: need clone here from body, sigh
```
Comment (pjenvey) [inline src/web/handlers.rs]: I don't think we need these 2 clones here any longer?

**Pull request merged**